homeassistant/components/lcn/helpers.py

```diff
@@ -1,18 +1,191 @@
 """Helpers for LCN component."""
+import logging
 import re
 
+import pypck
 import voluptuous as vol
 
-from homeassistant.const import CONF_NAME
+from homeassistant.const import (
+    CONF_ADDRESS,
+    CONF_DEVICES,
+    CONF_DOMAIN,
+    CONF_ENTITIES,
+    CONF_HOST,
+    CONF_IP_ADDRESS,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_UNIQUE_ID,
+    CONF_USERNAME,
+)
+
+from .const import (
+    CONF_ADDRESS_ID,
+    CONF_CONNECTIONS,
+    CONF_DIM_MODE,
+    CONF_DOMAIN_DATA,
+    CONF_HARDWARE_SERIAL,
+    CONF_HARDWARE_TYPE,
+    CONF_IS_GROUP,
+    CONF_RESOURCE,
+    CONF_SEGMENT_ID,
+    CONF_SK_NUM_TRIES,
+    CONF_SOFTWARE_SERIAL,
+    CONF_UNIQUE_DEVICE_ID,
+    CONNECTION,
+    DEFAULT_NAME,
+    DOMAIN,
+)
 
-from .const import DEFAULT_NAME
+_LOGGER = logging.getLogger(__name__)
 
 # Regex for address validation
 PATTERN_ADDRESS = re.compile(
     "^((?P<conn_id>\\w+)\\.)?s?(?P<seg_id>\\d+)\\.(?P<type>m|g)?(?P<id>\\d+)$"
 )
 
 
+DOMAIN_LOOKUP = {
+    "binary_sensors": "binary_sensor",
+    "climates": "climate",
+    "covers": "cover",
+    "lights": "light",
+    "scenes": "scene",
+    "sensors": "sensor",
+    "switches": "switch",
+}
+
+
+def get_device_config(unique_device_id, config_entry):
+    """Return the device configuration for given unique_device_id from ConfigEntry."""
+    for device_config in config_entry.data[CONF_DEVICES]:
+        if device_config[CONF_UNIQUE_ID] == unique_device_id:
+            return device_config
+    return None
+
+
+def get_device_address(device_config):
+    """Return a tuple with address information."""
+    return (
```

Comment:
My impression is that CONF_SEGMENT_ID, CONF_ADDRESS_ID, and CONF_IS_GROUP are only ever read in this method.  Would it work to just store the device address as a single device_config key to have fewer getters on the stack?  The motivation is trying to have fewer steps to understand what LcnServiceCall is doing.

Reply:
Ok, I will review the code and see if I can somehow unify address handling with the aim of removing this method. Might take some time.

Reply:
I was able to completely remove the unique_id for the devices in config_entry.data and replaced it by an address tuple (segment_id, address_id, is_group). This tuple is directly used at all other positions which demand for a LCN address.
Looks cleaner, now.